mixer/adapter/redisquota/redisquota.go

```diff
@@ -0,0 +1,400 @@
+		config.FIXED_WINDOW:   luaFixedWindow,
+		config.ROLLING_WINDOW: luaRollingWindow,
+	}
+)
+
+type (
+	builder struct {
+		quotaTypes    map[string]*quota.Type
+		adapterConfig *config.Params
+	}
+
+	handler struct {
+		// go-redis client
+		// connection pool with redis
+		client *redis.Client
+
+		// the limits we know about
+		limits map[string]*config.Params_Quota
+
+		// dimension hash map
+		dimensionHash map[*map[string]string]string
+
+		// list of algorithm LUA scripts
+		scripts map[config.Params_QuotaAlgorithm]*redis.Script
+
+		// indirection to support fast deterministic tests
+		getTime func() time.Time
+
+		// logger provided by the framework
+		logger adapter.Logger
+	}
+)
+
+///////////////// Configuration Methods ///////////////
+
+func (b *builder) SetQuotaTypes(quotaTypes map[string]*quota.Type) {
+	b.quotaTypes = quotaTypes
+}
+
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adapterConfig = cfg.(*config.Params)
+}
+
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	info := GetInfo()
+
+	if len(b.adapterConfig.Quotas) == 0 {
+		ce = ce.Appendf(info.Name, "quota should not be empty")
+	}
+
+	limits := make(map[string]*config.Params_Quota, len(b.adapterConfig.Quotas))
+	for idx := range b.adapterConfig.Quotas {
+		if len(b.adapterConfig.Quotas[idx].Name) == 0 {
+			ce = ce.Appendf(info.Name, "quotas.name should not be empty")
+			continue
+		}
+
+		limits[b.adapterConfig.Quotas[idx].Name] = &b.adapterConfig.Quotas[idx]
+
+		if b.adapterConfig.Quotas[idx].ValidDuration == 0 {
+			ce = ce.Appendf(info.Name, "quotas.valid_duration should be bigger must be > 0")
+			continue
+		}
+
+		if b.adapterConfig.Quotas[idx].RateLimitAlgorithm == config.ROLLING_WINDOW {
+			if b.adapterConfig.Quotas[idx].BucketDuration == 0 {
+				ce = ce.Appendf(info.Name, "quotas.bucket_duration should be > 0 for ROLLING_WINDOW algorithm")
+				continue
+			}
+
+			if b.adapterConfig.Quotas[idx].ValidDuration > 0 && b.adapterConfig.Quotas[idx].BucketDuration > 0 &&
+				b.adapterConfig.Quotas[idx].ValidDuration <= b.adapterConfig.Quotas[idx].BucketDuration {
+				ce = ce.Appendf(info.Name, "quotas.valid_duration: %v should be longer than quotas.bucket_duration: %v for ROLLING_WINDOW algorithm",
+					b.adapterConfig.Quotas[idx].ValidDuration, b.adapterConfig.Quotas[idx].BucketDuration)
+				continue
+			}
+		}
+
+		for index := range b.adapterConfig.Quotas[idx].Overrides {
+			if b.adapterConfig.Quotas[idx].Overrides[index].MaxAmount <= 0 {
+				ce = ce.Appendf(info.Name, "quotas.overrides.max_amount must be > 0")
+				continue
+			}
+
+			if len(b.adapterConfig.Quotas[idx].Overrides[index].Dimensions) == 0 {
+				ce = ce.Appendf(info.Name, "quotas.overrides.dimensions is empty")
+				continue
+			}
+
+			if _, err := getDimensionHash(b.adapterConfig.Quotas[idx].Overrides[index].Dimensions); err != nil {
+				ce = ce.Appendf(info.Name, "unable to initialize quota overrides dimensions",
```

Comment:
Consider suggesting to the user what they can do to fix this error.

Reply:
I added this condition to remove lint warning. But io.WriteString in the getDimensionHash never returns error. Removed checking error here